File: LLWallPaper.App/Services/BackendApiClient.cs
using System.Net.Http;
using System.Text.Json;
using LLWallPaper.App.Models;
using LLWallPaper.App.Utils;

namespace LLWallPaper.App.Services;

public sealed class BackendApiClient : IBackendApiClient
{
    private readonly HttpClient _httpClient;
    private readonly Func<string> _baseUrlProvider;
    private readonly AppLogger _logger;

    public BackendApiClient(HttpClient httpClient, Func<string> baseUrlProvider, AppLogger logger)
    {
        _httpClient = httpClient;
        _baseUrlProvider = baseUrlProvider;
        _logger = logger;
    }

    public async Task<IReadOnlyList<CardItem>> GetCardsAsync(CancellationToken cancellationToken)
    {
        var baseUrl = _baseUrlProvider()?.TrimEnd('/') ?? string.Empty;
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            throw new InvalidOperationException("Backend base URL is empty.");
        }

        try
        {
            var requestUrl = $"{baseUrl}/api/card-illustrations";
            using var response = await _httpClient.GetAsync(requestUrl, cancellationToken);
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException(
                    $"Backend returned {response.StatusCode} for {requestUrl}."
                );
            }

            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
            using var doc = await JsonDocument.ParseAsync(
                stream,
                cancellationToken: cancellationToken
            );
            if (doc.RootElement.ValueKind == JsonValueKind.Array)
            {
                return ParseArray(doc.RootElement, baseUrl);
            }

            if (
                doc.RootElement.TryGetProperty("cards", out var cards)
                && cards.ValueKind == JsonValueKind.Array
            )
            {
                return ParseArray(cards, baseUrl);
            }
        }
        catch (Exception ex)
        {
            _logger.Error("Failed to fetch cards from backend.", ex);
            throw;
        }
        return Array.Empty<CardItem>();
    }

    private static IReadOnlyList<CardItem> ParseArray(JsonElement array, string baseUrl)
    {
        var list = new List<CardItem>();
        foreach (var element in array.EnumerateArray())
        {
            if (!element.TryGetProperty("id", out var idProp))
            {
                continue;
            }

            var id = idProp.ValueKind switch
            {
                JsonValueKind.String => idProp.GetString() ?? string.Empty,
                JsonValueKind.Number => idProp.GetRawText(),
                _ => idProp.GetRawText(),
            };
            if (string.IsNullOrWhiteSpace(id))
            {
                continue;
            }

            var name = element.TryGetProperty("name", out var nameProp)
                ? nameProp.GetString() ?? id
                : id;
            var hasFull = HasAssetFlag(element, "full");
            if (!hasFull)
            {
                continue;
            }

            var imageUrl = $"{baseUrl}/api/card-illustrations/image/{id}?type=full";
            var thumbUrl = $"{baseUrl}/api/card-illustrations/image/{id}?type=half";

            list.Add(
                new CardItem
                {
                    Id = id,
                    Name = name,
                    ImageUrl = imageUrl,
                    ThumbnailUrl = thumbUrl,
                }
            );
        }

        return list;
    }

    private static bool HasAssetFlag(JsonElement element, string key)
    {
        if (!element.TryGetProperty("assets", out var assets))
        {
            return false;
        }

        if (!assets.TryGetProperty("images", out var images))
        {
            return false;
        }

        if (!images.TryGetProperty(key, out var flag))
        {
            return false;
        }

        return flag.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.String => bool.TryParse(flag.GetString(), out var parsed) && parsed,
            _ => false,
        };
    }
}

File: LLWallPaper.Mac/Sources/LLWallPaperMacCore/Infrastructure/BackendApiClient.swift
import Foundation

@MainActor
public final class BackendApiClient {
  private let session: URLSession
  private let logger: AppLogger

  public init(session: URLSession = .shared, logger: AppLogger) {
    self.session = session
    self.logger = logger
  }

  public func getCards(baseUrl: String) async throws -> [CardItem] {
    guard let baseUrl = URL(string: baseUrl.trimmingCharacters(in: .whitespacesAndNewlines)) else {
      throw BackendApiError.emptyBaseUrl
    }

    let requestUrl = baseUrl.appending(path: "api/card-illustrations")
    do {
      let (data, response) = try await session.data(from: requestUrl)
      guard let httpResponse = response as? HTTPURLResponse else {
        throw BackendApiError.invalidResponse
      }
      guard (200..<300).contains(httpResponse.statusCode) else {
        throw BackendApiError.httpStatus(httpResponse.statusCode)
      }

      return try Self.parseCards(data: data, baseUrl: baseUrl)
    } catch {
      logger.error("Failed to fetch cards from backend.", error)
      throw error
    }
  }

  public static func parseCards(data: Data, baseUrl: URL) throws -> [CardItem] {
    let json = try JSONSerialization.jsonObject(with: data)
    let array: [Any]

    if let rootArray = json as? [Any] {
      array = rootArray
    } else if let dictionary = json as? [String: Any], let cards = dictionary["cards"] as? [Any] {
      array = cards
    } else {
      return []
    }

    return array.compactMap { item in
      guard
        let dictionary = item as? [String: Any],
        let id = parseId(dictionary["id"]),
        hasImageAsset(dictionary, key: "full")
      else {
        return nil
      }

      let name = dictionary["name"] as? String ?? id
      let imageUrl = baseUrl.appending(path: "api/card-illustrations/image/\(id)")
        .appending(queryItems: [URLQueryItem(name: "type", value: "full")])
      let thumbnailUrl = baseUrl.appending(path: "api/card-illustrations/image/\(id)")
        .appending(queryItems: [URLQueryItem(name: "type", value: "half")])

      return CardItem(id: id, name: name, imageUrl: imageUrl, thumbnailUrl: thumbnailUrl)
    }
  }

  private static func parseId(_ value: Any?) -> String? {
    switch value {
    case let string as String:
      return string.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : string
    case let number as NSNumber:
      return number.stringValue
    default:
      return nil
    }
  }

  private static func hasImageAsset(_ dictionary: [String: Any], key: String) -> Bool {
    guard
      let assets = dictionary["assets"] as? [String: Any],
      let images = assets["images"] as? [String: Any],
      let value = images[key]
    else {
      return false
    }

    if let bool = value as? Bool {
      return bool
    }

    if let string = value as? String {
      return Bool(string) ?? false
    }

    return false
  }
}

public enum BackendApiError: LocalizedError, Equatable {
  case emptyBaseUrl
  case invalidResponse
  case httpStatus(Int)

  public var errorDescription: String? {
    switch self {
    case .emptyBaseUrl:
      return "Backend base URL is empty."
    case .invalidResponse:
      return "Backend returned an invalid response."
    case let .httpStatus(statusCode):
      return "Backend returned HTTP \(statusCode)."
    }
  }
}
